### PR TITLE
Move all common code about Trento tests to a lib

### PR DESCRIPTION
**Pull request by mpagot**
Better cypress.io integration by starting tests one by one
fail hook of the deploy_trento takes care to record all cluster logs

- Verification run: 
--- trento_jumphost http://1c242.qa.suse.de/tests/399  -  http://1c242.qa.suse.de/tests/439
--- trento http://1c242.qa.suse.de/tests/400  -  http://1c242.qa.suse.de/tests/441

### CODE REVIEW

**Comment by dzedro** [comment]
I had similar issue and should run `make tidy-full`, don't know if it will help, I didn't finish the PR.

**Comment by mpagot** [comment]
@dzedro  thanks, tested `make tidy-full` locally. Unfortunately does not seems to help. It produce no result and does not change any file. I also give a try to various `tools/tidy` script invocation.

**Comment by perlpunk** [comment]
I tried this PR locally and also did not get any changes from `make tidy-full`.
Unfortunately the output is not very informative. It would be good to know which version of Perl::Tidy it uses.

**Comment by perlpunk** [comment]
Ah. I just rebased the PR on the current master and now I get changes.
So try to rebase it =)

File: lib/trento.pm
Comment (alvarocarvajald) [inline lib/trento.pm:257]: Be careful with the terminology. If names in podman are the same as in docker, then `$container_name` is actually the **Image Name** and what you are passing to the command via `--name` is the **Container Name**. (I could be wrong though ... haven't used podman yet)

File: lib/trento.pm
Comment (alvarocarvajald) [inline lib/trento.pm:282]: `@arr` is not used in the function.

File: lib/trento.pm
Comment (alvarocarvajald) [inline lib/trento.pm:284]: Suggestion:
```
use File::Basename qw(basename);
...
my $test_filename = basename($_);
```

File: tests/sles4sap/trento/test_trento_deploy.pm
Comment (alvarocarvajald) [inline tests/sles4sap/trento/test_trento_deploy.pm:38]: Is this really needed?

**Review by alvarocarvajald — COMMENTED** [review]
Very nice PR. It addresses some of my questions from #14951 & #15019.

Thanks for adding POD documentation to `lib/trento.pm`. Once this is merged, it will be available in https://os-autoinst.github.io/os-autoinst-distri-opensuse/.

Did you check with `perldoc -U ./lib/trento.pm` or with `pod2man ./lib/trento.pm | nroff -man | less` for the formatting of the documentation? Documentation is there, and seems good enough in the output of `perldoc -U`, but if you compare it to the other libraries (for example `lib/utils.pm` or `lib/sles4sap.pm`) you'll see some differences. I suppose when it's rendered for the web page, it'll look better, but I am adding the comment here just in case you want to re-format for command line.

I have added more comments inline.

File: lib/trento.pm
Comment (mpagot) [inline lib/trento.pm:141]: code to be improved as it could result in teardown to fails http://1c242.qa.suse.de/tests/383#step/deploy_trento/28  fails as bash: kubectl: command not found  due to fact that teardown code try to collect cluster logs but the cluster is not there yet.

**Comment by mpagot** [comment]
> BTW, is this error expected?
> 
> http://1c242.qa.suse.de/tests/400#step/deploy_trento/52

Error is at http://1c242.qa.suse.de/tests/400#step/deploy_trento/50
```
[E:1][R] Trento Server start takes more then 3 minutes
         or not running at all. Exiting.
```

This failure does not depends on this perl test code

FYI @jankohoutek

File: lib/trento.pm
Comment (dzedro) [inline lib/trento.pm:41]: I don't understand this comment. :thinking:

File: lib/trento.pm
Comment (mpagot) [reply lib/trento.pm:44]: Probably inspired by https://peps.python.org/pep-0008/#blank-lines
Any suggestion for perl code in this context?

File: lib/trento.pm
Comment (mpagot) [reply lib/trento.pm:249]: I'll create a constant but probably not a variable. It is really just a constant and I do not see any additional value to let the "user" to configure it.

File: lib/trento.pm
Comment (mpagot) [reply lib/trento.pm:269]: not sure but it was after reading this definition : https://www.perlmonks.org/?node_id=685452 
what do you think?

File: tests/sles4sap/trento/test_trento_deploy.pm
Comment (mpagot) [reply tests/sles4sap/trento/test_trento_deploy.pm:38]: yes, I like to keep track of it. I was not able to  use a better curl arguments combination only available in a newer version. So I feel that I like to keep track of the running curl version.

File: lib/trento.pm
Comment (mpagot) [reply lib/trento.pm:41]: prefix is like **openqa-trento** with a dash in the middle. Most of the names are composed from this one. It is readable thanks to the dash.
Unfortunately the Azure ACR service does not support dash or underscore in the names, so only letter and number, that result in a less readable string like the one for **TRENTO_AZ_ACR_PREFIX**. The comment is a justification about the not so readable name. It is due to an Azure requirement.

File: lib/trento.pm
Comment (dzedro) [inline lib/trento.pm:40]: You don't need to export constant or ? I think it should be not done.
https://perldoc.perl.org/Exporter#What-Not-to-Export

File: lib/trento.pm
Comment (alvarocarvajald) [reply lib/trento.pm:61]: > Probably inspired by https://peps.python.org/pep-0008/#blank-lines Any suggestion for perl code in this context?

Oh, nice. Did not know about that convention for python. I am not aware of any sub convention for Perl, but looking some libraries at a glance inside os-autoinst-distri-opensuse I don't see anything similar.

File: lib/trento.pm
Comment (alvarocarvajald) [reply lib/trento.pm:322]: IMO, looking at that perlmonks.org article, this is a case of the call failing due to an issue within the method and not due to how it was called, so I would go with `die`, and that way you don't need to do `use Carp qw(croak)` either.

**Comment by alvarocarvajald** [comment]
> > Great PR! Please pay attention to the following items before merging:
> > **Files matching `lib/**.pm`:**
> > 
> > * [ ]   Consider adding or extending unit tests in t/
> > 
> > This is an automatically generated QA checklist based on modified files
> 
> Please @alvarocarvajald suggest to me some ideas about possible trento.pm unit tests please

Usually with libraries, one tries to write tests for all the methods defined in the library, but since this library has methods that perform installation and execution of podman images and containers, I am not really sure how to proceed. One idea is to look at `t/08-publiccloud.t` for some examples. It seems some of the tests there deal only with settings which is something you can do for the get* methods in `lib/trento.pm`

File: lib/trento.pm
Comment (dzedro) [reply lib/trento.pm:40]: I thought you don't need the constant out of `lib/trento.pm`
It's not explicitly bad, but it should be not done. You can do it if you like I think it's not first case of exported "variable/constant" :man_shrugging:

Comment (dzedro) [reply lib/trento.pm:40]: btw define constants on one place, there is another definition of VM_USER at line 123

Comment (perlpunk) [reply lib/trento.pm:40]: if the constant is not needed in a different file, don't export it.
And if you export something, prefer `@EXPORT_OK` over `@EXPORT` to only import on request.

File: lib/trento.pm
Comment (alvarocarvajald) [inline lib/trento.pm:218]: s/crypress/cypress/

File: lib/trento.pm
Comment (alvarocarvajald) [inline lib/trento.pm:171]: I think you can re-write this as `return unless ($machine_ip);` since there is no else in this if-block and you would remove a full 4-spaces indentation below as well as some brackets. It would improve readibility IMO.

File: lib/trento.pm
Comment (perlpunk) [inline lib/trento.pm:365]: double braces
```suggestion
        $self->cypress_log_upload('.txt', '.mp4');
```

File: tests/sles4sap/trento/deploy_trento.pm
Comment (perlpunk) [inline tests/sles4sap/trento/deploy_trento.pm:76]:
```suggestion
    $self->k8s_logs(qw(web runner));
```

File: tests/sles4sap/trento/setup_jumphost.pm
Comment (perlpunk) [inline tests/sles4sap/trento/setup_jumphost.pm:51]: That's a general suggestion:
```suggestion
    my $cypress_ver = get_var(TRENTO_CYPRESS_VERSION => '4.4.0');
```
If you use the "fat comma" `=>`, then you don't need to quote the string before.

File: tests/sles4sap/trento/test_trento_deploy.pm
Comment (perlpunk) [inline tests/sles4sap/trento/test_trento_deploy.pm:37]: Also a general suggestion:
```suggestion
    my $trento_url = "http://$machine_ip/";
```
I personally find interpolation more readable in many cases, where it's possible.
For longer concatenations I can also recommend to use [sprintf](https://perldoc.perl.org/functions/sprintf)

File: lib/trento.pm
Comment (mpagot) [reply lib/trento.pm]: There are some test files like **deploy_trento.pm** that are using many symbols from lib/trento.pm

Does not changing the overall list of exported symbols from EXPORT to EXPORT_OK result in a long and so difficult to maintain list of symbols in the test files?

Does it make sense to have two export lists, one using EXPORT and the other one using EXPORT_OK?

**Review by alvarocarvajald — APPROVED** [review]
LGTM

**Review by dzedro — APPROVED** [review]

File: lib/trento.pm
Comment (perlpunk) [inline lib/trento.pm:284]: I didn't see this earlier, but always leave a newline between pod and code. Some pod renderers will see this as malformed.

Comment (alvarocarvajald) [reply lib/trento.pm:284]: I didn't know about that, but I've always instinctively left empty lines surrounding all pod command as in https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/hacluster.pm or https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/sles4sap.pm (my main contributions to pod documentation in this repo).

@mpagot it seems it would be a good idea to apply Tina's suggestion even though most libraries within os-autoinst-distri-opensuse are not following it (examples: https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/susedistribution.pm, https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/opensusebasetest.pm or https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/utils.pm).

Comment (mpagot) [reply lib/trento.pm:284]: I play a little bit with it locally using `pod2html` exactly about this topic: "Where new line is needed"
For example new line is strictly needed after =head and =index

Ignoring for a second what is strictly needed for a proper rendering, and just focusing of the generic code readability: leaving one empty like between pod and code is maybe fine but somehow collide (personal opinion) with https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/15049#discussion_r895862427  (leaving 2 new lines between each method).

**Comment by alvarocarvajald** [comment]
I will merge this after Michele's last comment and the fact that some existing libraries in the repository do not leave an empty line between pod and code. I guess we should create a progress ticket to track the reformatting of the pod documentation in the libraries to apply Tina's suggestion.

**Comment by alvarocarvajald** [comment]
> I will merge this after Michele's last comment and the fact that some existing libraries in the repository do not leave an empty line between pod and code. I guess we should create a progress ticket to track the reformatting of the pod documentation in the libraries to apply Tina's suggestion.

I created https://progress.opensuse.org/issues/112544 to follow up on this.

**Pull request merged**
